### PR TITLE
Handles large amounts of data ingestion

### PR DESCRIPTION
**Pull request by umisora**
# BackGround
* When data integration is performed from a tenant with over 200 registered applications, an error occurs in the API call.

# What we did. 
* Implement a limit on the number of concurrent executions
* Implement a mechanism to reduce the number of target data

### CODE REVIEW

**Comment by umisora** [comment]
I tested with over 200 apps & 1 app have  over 2000 accounts.

File: src/integrate/integrate.ts [hunk 1]
```diff
@@ -42,14 +44,15 @@ const syncToAdmina = async (source: Source, inputs: Record<string, string>) => {
       console.log('Getting Azure AD data...')
       const azureAdData = await AzureAdSource.fetchApps(inputs)
       console.log('Registering custom service...')
-      await Promise.all(
-        azureAdData.map(async (app: AzureAdSource.AppInfo) => {
+      await PromisePool
```
Comment (peter-asztalos-mf) [inline src/integrate/integrate.ts:47]: PromisePool won't throw error, it collects all errors, so I think we manually have to handle it.

File: src/source/azuread.ts [hunk 2]
```diff
@@ -58,18 +66,31 @@ class AzureAD {
 
   private async getEnterpriseApplications(
     accessToken: string,
-    url = 'https://graph.microsoft.com/v1.0/applications'
+    serviceNames: string[] = [],
+    baseURL = 'https://graph.microsoft.com/v1.0/applications'
   ) {
+    const filterQueries = serviceNames.map(name => `(displayName eq '${name}')`).join(' or ');
+    const url = filterQueries ? `${baseURL}?$filter=${filterQueries}` : baseURL;
```
Comment (peter-asztalos-mf) [inline src/source/azuread.ts:73]: I'd use the URL class here.

```suggestion
    const url = new URL(baseURL);
    if (filterQueries) {
      url.searchParams.set('$filter', filterQueries);
    }
```

File: src/source/azuread.ts [hunk 3]
```diff
@@ -58,18 +66,31 @@ class AzureAD {
...
+    const isAppInfo = (obj: unknown): obj is AppInfo => {
+      return appInfoKeys.every(key => Object.prototype.hasOwnProperty.call(obj, key))
+    }
+    // PromisePoolで並列処理を実行
+    const { results } = await PromisePool
+      .for<AppInfo>(apps.filter(isAppInfo)) // 型ガードでフィルタリング
+      .withConcurrency(5) // 並列数を5に制限
+      .process(async (app: AppInfo) => {
         // Get the service principal id
         const servicePrincipalUrl = `https://graph.microsoft.com/v1.0/servicePrincipals?$filter=appId eq '${app.appId}'`
```
Comment (peter-asztalos-mf) [inline src/source/azuread.ts:95]: Same here. use URL.


```suggestion
        const servicePrincipalUrl = new URL('https://graph.microsoft.com/v1.0/servicePrincipals');
        servicePrincipalUrl.searchParams.set('$filter', `appId eq '${app.appId}'`)
```

File: src/integrate/integrate.ts (hunk 1 above)
Comment (umisora) [reply src/integrate/integrate.ts]: I fixed it 👏

File: src/source/azuread.ts (hunk 2 above)
Comment (umisora) [reply src/source/azuread.ts]: I fixed it 👏

File: src/source/azuread.ts (hunk 3 above)
Comment (umisora) [reply src/source/azuread.ts]: I fixed it 👏

**Comment by umisora** [comment]
@peter-asztalos-mf can you check again  please? 😸

File: src/integrate/integrate.ts [hunk 4]
```diff
@@ -44,15 +44,29 @@ const syncToAdmina = async (source: Source, inputs: Record<string, string>) => {
...
+            throw new Error(error.message);
+          }
         })
+      } catch (error) {
+        if (error instanceof Error) {
+          throw new Error(error.message);
+        } else {
+          throw new Error(`Failed to register application data. :${error}`);
+        }
+      }
```
Comment (peter-asztalos-mf) [inline src/integrate/integrate.ts:69]: I think this part is unnecessary.
You have a top lvl catch here already

https://github.com/moneyforward-i/admina-custom-service-action/blob/eecf724abbbd96b84e3bbe633bc1106a6a020f6a/src/index.ts#L44-L49

Also, Error contains message and stack, we only report the message, not the stack, which could be helpful with debugging, so in the top lvl try-catch I'd add an extra line before `if (error instanceof Error) {`

```
console.error('Failed due to error:', error);
```

File: src/integrate/integrate.ts [hunk 5]
```diff
@@ -44,15 +44,29 @@ const syncToAdmina = async (source: Source, inputs: Record<string, string>) => {
...
+              await AzureAdTransform.transformDataToAdmina(app),
+              inputs
+            )
+          })
+        errors.forEach(error => {
+          throw
+          if (error instanceof Error) {
+            throw new Error(error.message);
+          }
         })
```
Comment (peter-asztalos-mf) [inline src/integrate/integrate.ts:62]: I'd pint the all the errors and do a final throw at the end
```suggestion
        errors.forEach(error => {
          console.error('Sync failed', error);
        });
        if (errors.length > 0) {
          throw new Error("The syncing process couldn't finish completely, please check the errors");
        }
```

File: src/source/azuread.ts [hunk 6]
```diff
@@ -87,31 +89,42 @@ class AzureAD {
...
+        if (error instanceof Error) {
+          throw new Error(error.message);
         }
-        return appInfo
       })
-    let appInfos = results
+      appInfos = results
+    } catch (error) {
+      throw new Error(`Fail to fetch applications from AzureAd: ${error}`);
+    }
```
Comment (peter-asztalos-mf) [inline src/source/azuread.ts:127]: same as in other place, try-catch not needed and print messages and do fnal throw

File: src/source/azuread.ts [hunk 7]
```diff
@@ -152,23 +165,33 @@ class AzureAD {
...
-    let userInfos = results
+      userInfos = results
+
+    } catch (error) {
+      if (error instanceof Error) {
+        throw new Error(error.message);
+      } else {
+        throw new Error(`Failed to register application data. :${error}`);
+      }
+    }
```
Comment (peter-asztalos-mf) [inline src/source/azuread.ts:195]: Same. (btw, this part of code doesn't handle errors at all)

File: src/source/azuread.ts [hunk 8]
```diff
@@ -211,17 +234,28 @@ class AzureAD {
...
+          throw new Error(error.message);
+        })
+        filteredResults = results.filter(appInfo => appInfo !== null) as AppInfo[]
+      } catch (error) {
+        if (error instanceof Error) {
+          throw new Error(error.message);
+        } else {
+          throw new Error(`Failed to fetch list of applications. :${error}`);
+        }
+      }
```
Comment (peter-asztalos-mf) [inline src/source/azuread.ts:258]: Same

**Comment by umisora** [comment]
This PR has been merged into the following PR.
#5

**Pull request closed without merging**